testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -0,0 +1,927 @@
+  //   multiple requests/responses (verifying contents received)
+  //   server transport shutdown triggers client shutdown (via GOAWAY)
+  //   queued message InputStreams are closed on stream cancel
+  //     (and maybe exceptions handled)
+
+  @Test
+  public void serverNotListening() {
+    server = null;
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    verify(mockClientTransportListener, timeout(TIMEOUT_MS)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(statusCaptor.capture());
+    assertCodeEquals(Status.UNAVAILABLE, statusCaptor.getValue());
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    verify(mockClientTransportListener, never()).transportReady();
+  }
+
+  @Test
+  public void clientStartStop() throws Exception {
+    server.start(serverListener);
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    client.shutdown();
+    verify(mockClientTransportListener, timeout(TIMEOUT_MS)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(any(Status.class));
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+  }
+
+  @Test
+  public void clientStartAndStopOnceConnected() throws Exception {
+    server.start(serverListener);
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    verify(mockClientTransportListener, timeout(TIMEOUT_MS)).transportReady();
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    client.shutdown();
+    verify(mockClientTransportListener, timeout(TIMEOUT_MS)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(any(Status.class));
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    assertTrue(serverTransportListener.waitForTermination(TIMEOUT_MS, TimeUnit.MILLISECONDS));
+    server.shutdown();
+    assertTrue(serverListener.waitForShutdown(TIMEOUT_MS, TimeUnit.MILLISECONDS));
+    server = null;
+  }
+
+  @Test
+  public void serverAlreadyListening() throws Exception {
+    client = null;
+    server.start(serverListener);
+    Server server2 = newServer();
+    thrown.expect(IOException.class);
+    server2.start(new MockServerListener());
+  }
+
+  @Test
+  public void openStreamPreventsTermination() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    clientStream.start(mockClientStreamListener);
```

Comment:
There is a race between transport shutdown and stream start. Shall we test that an unstarted stream would also prevent termination?